### ISSUE TITLE
RC1 @angular bundling issue in order to reduce hundreds of request

### ISSUE THREAD

**Issue by oodboo** [angular/angular#8717]
I am not sure whether to ask this question in webpack or @angular repo, as there was https://angular.io/docs/ts/latest/guide/webpack.html tutorial which did not help me, and as I have mainly issue with RC1 @angular 400+ requests I decided that maybe best place is to ask experts here.

Although this question/issue has similiarties with the #8591 and #8542 there are not the same, and I should mention that I have tried some of the things mentioned there but then I have discovered another issue with angular-cli (angular/angular-cli/issues/828)

I am trying to do is following:

Currently I am running node (expressjs)+jade as MVC server side project, (where some pages are served without angular, home login etc.).
Inside of the same project there is a dashboard that runs angular2 (rc1 on client) project. 

Structure roughly looks like this:
```
+ bin 
      - www
+ controllers (routes-expressJs)
      - home.js
+ public 
     + css
     + js
         -+dist (angular2 build)
             --(all the js and angular2 template project  files )
+ views
    + login
    + home 
         + angular.html
```

When page is requested it is intercepted by express.js that renders page from angular.html template.
At this point all the angular client .js files are called.

Now issue with this is that at this point I will get 400+ requests, which will happen with basic Quick starter project as well, as RC1 files are not bundled.

As I have server side and client side running in the same project, I would like to use "nodemon ./bin/www -e js,html,css,tsc" to monitor my changes.

I have tried running webpack but it will bundle all files into one or can create separate files but still everything will be bundled.

What I ideally want is to run bundling vendor and polyfill files only once, and then manually include/reference them in index.html (angular.html) file.
From there I would import "src/main.js" (angular boot) by System.Import() in usual way while everything would be compiled by a background running "tsc -w" process: 

Index.html code below:

```
    <script type="text/javascript" src="../../dist/polyfills.3b18806f261793ecd4aa.js"></script>
    <script type="text/javascript" src="../../dist/vendor.3b18806f261793ecd4aa.js"></script>
    <script>
        System.import('app/main');
    </script>
```
Issue with this approach is that all libraries withing Webpack are contained and nothing is visible outside, I am not sure how to make them visible so I can use them by SystemJs.

There are multiple reason I need this alignment, also I have noticed there is an overhead for tracking changes in Webpack even smallest change need between 5-15 seconds to rebuild.

Now this is not just about angular, but ideally I would like to bundle other libraries rxjs, etc ...

Does anyone knows if not with Webpack, is there any tool that will allow me to do this?

**Comment by antonybudianto7** [comment]
[SystemJS Builder](https://github.com/systemjs/builder), I'm using it for my [starter](https://github.com/antonybudianto/angular2-starter) about 8 months and still going good!

**Comment by oodboo** [comment]
@antonybudianto7 
This specific problem is with version rc0,1 before I have not had issue with 400 request as with angular2 I was using npm install beta.-0-17 ... and inside there was a bundled file.

I am fine with SystemJS, but it will not fetch anything from bundle that is processed with WebPack.

Only difference I can see is that in your starter  SystemJS config file you are referencing a "node_module" path,  and in your gulpfile.js (+gulp.config.js) I cannot see that you anywhere **bundling** any @angular .js files ... please could you explain ...

**Comment by antonybudianto7** [comment]
it's [here](https://github.com/antonybudianto/angular2-starter/blob/master/tasks/build.js#L24)

**Comment by oodboo** [comment]
Thank you @antonybudianto7 now I see, you have entire env. setup there, I think this will do the trick, I will need to adjust that "serve-build" to become one-off asset build for DEV ...

**Comment by oodboo** [comment]
@antonybudianto7 I have spent last 4hrs trying to make this work, with your example and this is not what I need. This is exactly what WebPack does.
You are bundling everything into bundle file app + libs and I do not want that , what I want is following:

```
>> bundle all packages in single file 
        @angular/common
        @angular/compiler
        @angular/core
        @angular/http
        @angular/platform-browser
        @angular/platform-browser-dynamic
        @angular/router
        @angular/testing
        lodash
        rxjs
       |----------------------------> lib-99999999.js (bundled vendor files)

then in html file:

  <body>
    <as-main-app>
      <i class="center-fix main-spinner fa fa-spin fa-spinner"></i>
    </as-main-app>

    <script src="build/assets/lib-99999999.js"></script>
    
    <script src="node_modules/systemjs/dist/system.js"></script>
    <script src="systemjs.conf.js"></script>
    <script>
      System.import('app/main.js')
        .catch(function(err) { console.error(err); });
    </script>
  </body>
```

Now somehow I need to config SystemJS to say that all the packages are contained in single file:
**lib-99999999.js**

But I am not sure that it is possible to do something like that?

**Comment by oodboo** [comment]
For now the closest thing I could come up is this, it uses what is already provided in bundle, it will significantly reduce number of requests:
```
  <body>
    <my-app></my-app>

    <script src="node_modules/jquery/dist/jquery.js"></script>
    <script src="node_modules/bootstrap-sass/assets/javascripts/bootstrap.js"></script>

    <!-- 1. Load libraries -->
    <script src="node_modules/zone.js/dist/zone.js"></script>
    <script src="node_modules/reflect-metadata/Reflect.js"></script>

    <!-- 2. Polyfill(s) for older browsers -->
    <script src="node_modules/es6-shim/es6-shim.min.js"></script>

    <script src="node_modules/systemjs/dist/system.js"></script>
    <!-- 3 Rxjs load as I could not find the way to do it with SystemJs -->
    <script src="node_modules/rxjs/bundles/Rx.min.js"></script>
    
    <script src="systemjs.conf.js"></script>
    <script>
      System.import('app/main.js')
        .catch(function(err) { console.error(err); });
    </script>
  </body>

```
and systemjs.conf.js file:
```
(function(global) {
    global.ENV = 'development'

    var paths = {
        'n:*': 'node_modules/*'
    };
    var map = {
        'app': 'app',
        // 'rxjs': 'n:rxjs',
        '@angular': 'n:@angular',
        'lodash': 'n:lodash'
    };

    var packages = {
        'app': {
            defaultExtension: 'js'
        },
        'lodash' : {
            main: 'lodash.js',
            defaultExtension: 'js'
        },
    };

    var packageNames = [
        '@angular/common',
        '@angular/compiler',
        '@angular/core',
        '@angular/http',
        '@angular/platform-browser',
        '@angular/platform-browser-dynamic',
        '@angular/router',
        '@angular/testing',
    ];

    packageNames.forEach(function(pkgName) {
        var name = pkgName.indexOf("/")>0 ? pkgName.split("/")[1] : pkgName; 
        console.log("Name:", name);
        packages[pkgName] = { main: name + '.umd.js', defaultExtension: 'js' };
    });
    var config = {
        map: map,
        packages: packages,
        paths: paths
    };
    if (global.filterSystemConfig) { global.filterSystemConfig(config); }
    System.config(config);
})(this);
```
If anyone has better way please sugest?

**Comment by radusuciu** [comment]
@oodboo This makes it more bearable. Thank you.

**Comment by antonybudianto7** [comment]
@oodboo Sorry, our approach is different, I never intended to make systemjs coupled to my production bundle, so I remove the config and systemjs ref when bundling, that's why my final `index.html`, it didn't have something like
```html
<script src="systemjs.conf.js"></script>
    <script>
      System.import('app/main.js')
        .catch(function(err) { console.error(err); });
    </script>
```
because i already have self executing bundle

**Comment by stewhouston** [comment]
I was able to create a bundling task with gulp using [systemjs-builder](https://github.com/systemjs/builder). 

I am skeptical of using this package as an npm module, but I based my task off of [https://github.com/ngstarter/ngstarter-systemjs-tasks](https://github.com/ngstarter/ngstarter-systemjs-tasks).

I am using an angular 2 app based off of the structure and system-config.ts generated by [angular-cli](https://github.com/angular/angular-cli). 

Posting because it took me several days to figure out how to go about this. Example bundling task with gulp [here](https://gist.github.com/stewhouston/27cac18c64d03b14de8b39389cbaa1c5). I then replaced the System.import command in the angular2 'index.html' file with a script tag pointing to the generated bundle file. Note that I had to paste some of the custom node_modules (e.g. [angular2-cookie](https://www.npmjs.com/package/angular2-cookie)) into the 'dist' directory as the builder could not find them otherwise.

Hopefully this process is available through the core angular2 tools eventually.

**Comment by IgorMinar** [comment]
the systemjs.config.js in https://angular.io/docs/ts/latest/quickstart.html shows two kinds of configurations - one where all individual files are fetched and the other where umd bundles are used.

it is however better to bundle the files yourself, but in that case you shouldn't mix SystemJS and webpack. Instead configure webpack to create two bundles - the vendor bundle and that app bundle as suggested by others above.

 the angular-cli already does this kind of bundling for you, so if you use it you don't need to setup anything yourself.

**Comment by LukeNZ** [comment]
@IgorMinar Is there any chance the documentation could be updated to provide more details on the different bundling strategies available out of the box, especially relating to systemjs? It's definitely lacking in this area, and not all of us have the time to mess around with webpack. 

Right now due to #9431 (router v3) and #9359 (rxjs), it is only possible to bundle a subset of the total number of dependencies that angular 2 uses.

We're down to ~450 requests from ~750, but this is an area that needs improvement IMO.
